/**
 * This file contains METAL definitions for the macros and functions needed for the
 * common compute framework.
 */

uint3 threadIdx [[thread_position_in_threadgroup]];
uint3 blockIdx [[threadgroup_position_in_grid]];
uint3 blockDim [[threads_per_threadgroup]];
uint3 gridDim [[threadgroups_per_grid]];

#define __global__ kernel
#define __device__
#define __shared__ threadgroup
#define __syncthreads() threadgroup_barrier(mem_flags::mem_threadgroup | mem_flags::mem_device)
#define __threadfence() atomic_thread_fence(mem_flags::mem_device, memory_order_seq_cst, thread_scope_device)
#define __launch_bounds__(x)
#define __clz clz
#define __popc popcount
#define __ffs(x) ((x) == 0 ? 0 : ctz(x)+1)
#define __float_as_uint(x) as_type<uint>(x)
#define __half half
#define __half2float(x) float(x)

#define KERNEL extern "C" __global__
#define DEVICE __device__
#define LOCAL __shared__
#define LOCAL_ARG threadgroup
#define GLOBAL device
#define PRIVATE thread
#define RESTRICT __restrict__
#define LOCAL_ID threadIdx.x
#define LOCAL_SIZE blockDim.x
#define GLOBAL_ID (blockIdx.x*blockDim.x+threadIdx.x)
#define GLOBAL_SIZE (blockDim.x*gridDim.x)
#define GROUP_ID blockIdx.x
#define NUM_GROUPS gridDim.x
#define SYNC_THREADS __syncthreads();
#define SYNC_WARPS simdgroup_barrier(mem_flags::mem_threadgroup);
#define MEM_FENCE
#define ATOMIC_ADD(dest, value) atomicAdd(dest, value)

typedef long mm_long;
typedef unsigned long mm_ulong;

inline int atomicAdd(device int* dest, int value) {
    return atomic_fetch_add_explicit((device atomic_int*) dest, value, memory_order_relaxed);
}

inline unsigned int atomicAdd(device unsigned int* dest, unsigned int value) {
    return atomic_fetch_add_explicit((device atomic_uint*) dest, value, memory_order_relaxed);
}

inline float atomicAdd(device float* dest, float value) {
    return atomic_fetch_add_explicit((device atomic_float*) dest, value, memory_order_relaxed);
}

inline unsigned long atomicAdd(device unsigned long* dest, unsigned long value) {
    // Probe only, wrong forces: a plain read-modify-write, no atomic.
    *dest += value;
    return 0;
}

inline float atomicMin(device float* dest, float value) {
    return as_type<float>(atomic_fetch_min_explicit((device atomic_uint*) dest, as_type<uint>(value), memory_order_relaxed));
}

inline float atomicMax(device float* dest, float value) {
    return as_type<float>(atomic_fetch_max_explicit((device atomic_uint*) dest, as_type<uint>(value), memory_order_relaxed));
}

inline half __float2half_ru(float x) {
    half h = half(x);
    return (float(h) < x ? nextafter(h, half(INFINITY)) : h);
}

#define make_short3 short3
#define make_int2 int2
#define make_int3 int3
#define make_int4 int4
#define make_float2 float2
#define make_float3 float3
#define make_float4 float4

#define sqrtf sqrt
#define expf exp
#define logf log
#define cosf cos
#define sinf sin
#define tanf tan
#define acosf acos
#define asinf asin
#define atanf atan
#define atan2f atan2
#define powf pow
#define erff erf
#define erfcf erfc
#define fmaf fma
#define fabsf fabs
#define truncf trunc

#define __fsqrt_rn precise::sqrt
#define __frsqrt_rn fast::rsqrt
#define __expf fast::exp
#define __logf fast::log

inline float erfc(float x) {
    float t = fabs(x);
    float u = 1.0f/(1.0f+0.47f*t);
    float poly = ((((((0.0966678f*u-0.4607003f)*u+0.6896449f)*u-0.21611532f)*u+0.3931878f)*u+0.22740916f)*u+0.2701903f)*u-0.00028434425f;
    float result = (t > 9.0f ? 0.0f : poly*exp(-t*t));
    return (x < 0.0f ? 2.0f-result : result);
}

inline float erf(float x) {
    if (fabs(x) >= 0.5f)
        return 1.0f-erfc(x);
    float x2 = x*x;
    return 1.1283791670955126f*x*(1.0f+x2*(-1.0f/3.0f+x2*(1.0f/10.0f+x2*(-1.0f/42.0f+x2*(1.0f/216.0f+x2*(-1.0f/1320.0f))))));
}

__device__ inline long realToFixedPoint(float x) {
    // Faster way to calculate static_cast<long long>(x * 0x100000000) with exactly the same
    // results but less instructions.
    float integral = truncf(x);
    float fractional = (x - integral) * 0x100000000;
    unsigned int integral_u32 = static_cast<int>(integral);
    unsigned int fractional_u32 = static_cast<unsigned int>(fabsf(fractional));
    // A negative real number (with non-zero fractional) needs rounding-down x for integral and
    // changing fractional's sign. However, -1 is used as a threshold instead of 0 because, when
    // fractional is in (-1; 0], fractional_u32 is 0 and the number is considered an integer.
    bool isNegReal = fractional <= -1.0f;
    return (static_cast<unsigned long>(isNegReal ? integral_u32 - 1 : integral_u32) << 32) |
            static_cast<unsigned long>(isNegReal ? 0 - fractional_u32 : fractional_u32);
}
